Index: coregrind/m_syswrap/syswrap-generic.c
===================================================================
--- coregrind/m_syswrap/syswrap-generic.c	(revision 13886)
+++ coregrind/m_syswrap/syswrap-generic.c	(working copy)
@@ -2157,7 +2157,17 @@
                                        arg4 | VKI_MAP_FIXED,
                                        arg5, arg6);
    }
-
+   
+   /* Another refinement: a hack to work around the hugepages
+    * For some reason huge page allocation has a problem with 
+    * MAP_FIXED, so ignore it. */
+   if(sr_isError(sres)){
+     /* try again, MAP_FIXED */
+     sres = VG_(am_do_mmap_NO_NOTIFY)(
+              advised, arg2, arg3, arg4,
+              arg5, arg6);
+   }
+   
    if (!sr_isError(sres)) {
       ULong di_handle;
       /* Notify aspacem. */
